src/transaction/lock_table.c

```diff
@@ -117,75 +123,81 @@ LOCK_COMPATIBILITY lock_Comp[11][11] = {
  *
- *    NULL | N/A     N/A   NULL  SCH-S     IS      S     IX    SIX      U      X  SCH-M
+ *    NULL | N/A     N/A   NULL  SCH-S     IS      S     IX    SIX      U      X  SCH-M        BU
+ *                                                                                             
+ *   SCH-S | N/A     N/A  SCH-S  SCH-S     IS      S     IX    SIX    N/A      X  SCH-M         X
  *
- *   SCH-S | N/A     N/A  SCH-S  SCH-S     IS      S     IX    SIX    N/A      X  SCH-M
+ *      IS | N/A     N/A     IS     IS     IS      S     IX    SIX    N/A      X  SCH-M        IS
  *
- *      IS | N/A     N/A     IS     IS     IS      S     IX    SIX    N/A      X  SCH-M
+ *       S | N/A     N/A      S      S      S      S    SIX    SIX      U      X  SCH-M         S
  *
- *       S | N/A     N/A      S      S      S      S    SIX    SIX      U      X  SCH-M
+ *      IX | N/A     N/A     IX     IX     IX    SIX     IX    SIX    N/A      X  SCH-M        IX
  *
- *      IX | N/A     N/A     IX     IX     IX    SIX     IX    SIX    N/A      X  SCH-M
+ *     SIX | N/A     N/A    SIX    SIX    SIX    SIX    SIX    SIX    N/A      X  SCH-M       SIX
  *
- *     SIX | N/A     N/A    SIX    SIX    SIX    SIX    SIX    SIX    N/A      X  SCH-M
+ *       U | N/A     N/A      U    N/A    N/A      U    N/A    N/A      U      X    N/A       N/A
  *
- *       U | N/A     N/A      U    N/A    N/A      U    N/A    N/A      U      X    N/A
+ *       X | N/A     N/A      X      X      X      X      X      X      X      X  SCH-M         X
  *
- *       X | N/A     N/A      X      X      X      X      X      X      X      X  SCH-M
+ *   SCH-M | N/A     N/A  SCH-M  SCH-M  SCH-M  SCH-M  SCH-M  SCH-M    N/A  SCH-M  SCH-M     SCH-M
  *
- *   SCH-M | N/A     N/A  SCH-M  SCH-M  SCH-M  SCH-M  SCH-M  SCH-M    N/A  SCH-M  SCH-M
+ * BU_LOCK | N/A     N/A     BU      X     IS      S     IX    SIX    N/A      X  SCH-M        BU
  * ------------------------------------------------------------------------------------------------
  * N/A : not applicable
  */
 
-LOCK lock_Conv[11][11] = {
+LOCK lock_Conv[12][12] = {
   /* N/A */
   {NA_LOCK, NA_LOCK, NA_LOCK, NA_LOCK, NA_LOCK, NA_LOCK, NA_LOCK, NA_LOCK,
-   NA_LOCK, NA_LOCK, NA_LOCK}
+   NA_LOCK, NA_LOCK, NA_LOCK, NA_LOCK}
   ,
   /* NON2PL */
   {NA_LOCK, NA_LOCK, NA_LOCK, NA_LOCK, NA_LOCK, NA_LOCK, NA_LOCK, NA_LOCK,
-   NA_LOCK, NA_LOCK, NA_LOCK}
+   NA_LOCK, NA_LOCK, NA_LOCK, NA_LOCK}
   ,
   /* NULL */
   {NA_LOCK, NA_LOCK, NULL_LOCK, SCH_S_LOCK, IS_LOCK, S_LOCK, IX_LOCK,
-   SIX_LOCK, U_LOCK, X_LOCK, SCH_M_LOCK}
+   SIX_LOCK, U_LOCK, X_LOCK, SCH_M_LOCK, BU_LOCK}
   ,
   /* SCH-S */
   {NA_LOCK, NA_LOCK, SCH_S_LOCK, SCH_S_LOCK, IS_LOCK, S_LOCK, IX_LOCK,
-   SIX_LOCK, NA_LOCK, X_LOCK, SCH_M_LOCK}
+   SIX_LOCK, NA_LOCK, X_LOCK, SCH_M_LOCK, X_LOCK}
   ,
   /* IS */
   {NA_LOCK, NA_LOCK, IS_LOCK, IS_LOCK, IS_LOCK, S_LOCK, IX_LOCK, SIX_LOCK,
-   NA_LOCK, X_LOCK, SCH_M_LOCK}
+   NA_LOCK, X_LOCK, SCH_M_LOCK, IS_LOCK}
```

Comment:
I don't think IS + BU can become IS; the resulting lock cannot be less restrictive than any of the locks. IS is less restrictive than BU.

Since BU is only compatible to BU and SCH_S, combining with any lock in the IS->X range could only be less compatible. SCH_S compatibility stands, but BU is not compatible with that range; so resulting lock should be compatible only with SCH_S, which means X_LOCK.

Reply:
I think IS + BU = BU. 
When a reader wants to promote IS to BU, it should be BU. 

difficult case is (BU + IS). 
When it is BU, a loader may read the table. 
When it is X, a loader might be blocked by other loaders to read the table.